### ISSUE TITLE
Adding support for `eslint.config.mjs` and `eslint.config.cjs` files.

### ISSUE THREAD

**Issue by yittoo** [microsoft/vscode-eslint#1707]
I am not sure if it is something that you do want to do until FlatConfig is final. But for cases where we want to use `eslint.config.mjs` we are adding a workspace overwrite as follows. I think this could be extended in the extension source code.

```
"eslint.options": {
  "overrideConfigFile": "eslint.config.mjs"
}
```

I can help with the code adjustment if this feature is desired. Cheers.

**Comment by dbaeumer** [comment]
How does that work when using ESLint on the command line?

**Comment by yittoo** [comment]
@dbaeumer with `eslint -c eslint.config.mjs` cli flag

**Comment by dbaeumer** [comment]
Since this is a command line option you need to use `eslint.options` to point to a different config file than the default one. I am reluctant to add any code to detect and choose config files besides that standard behavior of eslint npm package.

I will close the issue since this is the expected behavior.

**Comment by itpropro** [comment]
@dbaeumer  `eslint.config.js` is considered the new standard and the `.eslint` options as legacy reference. I would definitely see it as a bug in this VSCode extension, if the default way of writing eslint configs would not be detected by default.

**Comment by dbaeumer** [comment]
The extension detects by default anything that the `eslint` npm module (https://www.npmjs.com/package/eslint) uses as default. I don't have any special code that detects the config file. This is why I said if you need to pass `-c` on the command line you need to let the extension know the same via `eslint.options`

**Comment by jamielaw-jellyfish** [comment]
I'm encountering a related issue.

Steps to reproduce:

Start a new blank project with the following `package.json`
```json
{
  "scripts": {
    "lint": "eslint ."
  },
  "devDependencies": {
    "eslint": "^8.51.0"
  },
  "type": "module"
}
```

`eslint.config.js`

```js
export default [
  {
    rules: {
      semi: "error",
    }
  }
];
```

and `index.js`

```js
const foo = 42
```

VS Code gives no red underlines inside of `index.js` but running `npm run lint` will output the following errors in the console:

```
/Users/jamielaw/projects/linting/index.js
  1:15  error  Missing semicolon  semi

✖ 1 problem (1 error, 0 warnings)
  1 error and 0 warnings potentially fixable with the `--fix` option.
```

Switching `eslint.config.js` to the following `.eslintrc.js`

```js
module.exports = {
  env: {
    es6: true
  },
  rules: {
    semi: 'error'
  }
};
```

and VS Code will highlight code errors automatically as expected. So, ESLint seems to understand the new config file format, but VS Code isn't picking it up.

**Comment by yittoo** [comment]
@jamielaw-jellyfish  If i am not mistaken this feature is experimental still (or perhaps recently moved to stable?). Maybe the following will help fix it.

`.vscode/settings.json`
```json
{
  "eslint.experimental.useFlatConfig": true,
}
```

**Comment by jamielaw-jellyfish** [comment]
@yittoo It did not work in this case, at least not for me.

However, just knowing that in time it will be available is good enough. For the time being, I will keep using the old config format, but keep an eye out for when support for the new format becomes publicly available. :-)

**Comment by dbaeumer** [comment]
@jamielaw-jellyfish can you please provide me with a Github repository I can clone that demos this. This ensures we are looking at the same problem since things like this depend on many variables.

**Comment by jamielaw-jellyfish** [comment]
@dbaeumer Here you go:

https://github.com/jamielaw-jellyfish/eslint-config-test

**Comment by dbaeumer** [comment]
Could it be that you have something globally installed. Running `npm run lint` fails for me like this:

<img width="1141" alt="image" src="https://github.com/microsoft/vscode-eslint/assets/1931590/7b00cd7f-a956-42fb-9271-3b73f4f06413">

**Comment by jamielaw-jellyfish** [comment]
I realise the error I made in the repo: `.eslint.config.js` should not have a leading dot. Remove that and the command line should find the config file.

**Comment by dbaeumer** [comment]
Works for me as expected in VS Code

<img width="809" alt="image" src="https://github.com/microsoft/vscode-eslint/assets/1931590/aff9b34f-74c8-487b-a97d-b9592b7e092d">


Anything special in the ESLint otuput log?

**Comment by jamielaw-jellyfish** [comment]
So, I have been able to get it working, I think. Here was my process:

1. I updated VS Code. I was on 1.83.0 before and now I am on 1.83.1
2. I restarted the ESLInt Server. Without `.vscode/settings.json` I got errors in the ESLint output log that no config file could be found. Adding the file was enough to get it working.

So, I think the solution was to add the file, but also to restart ESLint and/or VS Code.

Sorry for taking you down this rabbit hole.

**Comment by dbaeumer** [comment]
No problem. Happy to hear you could solve it.

**Comment by EightArmCode** [comment]
I never was able to get `.mjs` working on my system.

**Comment by dbaeumer** [comment]
@EightArmCode can you please provide me with a GitHub repository I can clone that contains a minimal setup to reproduce the problem.

**Comment by Balvajs** [comment]
I am facing the same problem. I have a monorepo where each library has `eslint.config.mjs`, which works perfectly in CLI, however, VSCode plugin is not able to pick that up. If there would be some VSCode config that would force the plugin to "auto-discover" `eslint.config.mjs` instead of `eslint.config.js` I would be happy to use it 🤞 (or ideally if it would support all 3 formats: `.js`, `.mjs`, `.cjs`)

**Comment by mattc-solutions** [comment]
Want to share the combination of https://github.com/microsoft/vscode-eslint/issues/1707#issue-1894673734 and https://github.com/microsoft/vscode-eslint/issues/1707#issuecomment-1752773109 that got things working for me:

```json
{
  "eslint.experimental.useFlatConfig": true,
  "eslint.options": {
    "overrideConfigFile": "./eslint.config.mjs"
  }
}
```

**Comment by bmakan** [comment]
Something is definitely wrong. I keep getting the `Preprocessing error: No ESLint configuration found in /home/bmakan/ws/myapp/src/client` no matter what settings I use.

My project root is at `/home/bmakan/ws/myapp`. So I set working directories:

```json
    "eslint.workingDirectories": [
        "./src/client",
    ],
```

I tried various other settings, `mjs` extension, `type: module` in package.json, etc. Nothing worked.
